source/wayland/display.c

```diff
@@ -526,6 +526,100 @@ static const struct wl_keyboard_listener wayland_keyboard_listener = {
     .repeat_info = wayland_keyboard_repeat_info,
 };
 
+static void wayland_touch_down(void *data, struct wl_touch *wl_touch,
+    uint32_t serial, uint32_t time, struct wl_surface *wl_surface,
+    int32_t id, wl_fixed_t surface_x, wl_fixed_t surface_y) {
+  wayland_seat *self = data;
+  if (id >= MAX_TOUCHPOINTS) {
+    return;
+  }
+  self->touches[id].x = wl_fixed_to_int(surface_x);
+  self->touches[id].start_y = self->touches[id].move_y = wl_fixed_to_int(surface_y);
+  self->touches[id].start_time = self->touches[id].move_time = time;
+  RofiViewState *state = rofi_view_get_active();
+
+  if (state == NULL) {
+    return;
+  }
+  rofi_view_handle_mouse_motion(state, self->touches[id].x, self->touches[id].start_y,
+                                FALSE);
+}
+
+static void wayland_touch_up(void *data, struct wl_touch *wl_touch,
+    uint32_t serial, uint32_t time, int32_t id) {
+  wayland_seat *self = data;
+  if (id >= MAX_TOUCHPOINTS) {
+    return;
+  }
+  gboolean has_moved = self->touches[id].start_time != self->touches[id].move_time;
+  if (has_moved) {
+    return;
+  }
+  RofiViewState *state = rofi_view_get_active();
+
+  if (state == NULL) {
+    return;
+  }
+  int key = KEY_ENTER;
+  if (time - self->touches[id].start_time > 200) {
+    key = KEY_ESC;
+  }
+  nk_bindings_seat_handle_key(wayland->bindings_seat, NULL,
```

Comment:
on some rofi launches, this doesn't work (i.e. if it doesn't work, it won't magically fix itself if you click multiple times). TODO figure out why